File: src/server/core/go/projection.go
/* src/server/core/go/projection.go */

package seam

// applyProjection prunes loader data to only projected fields.
// nil projections = keep all data.
func applyProjection(data map[string]any, projections map[string][]string) map[string]any {
	if len(projections) == 0 {
		return data
	}

	result := make(map[string]any, len(data))
	for key, value := range data {
		fields, hasProjection := projections[key]
		if !hasProjection {
			result[key] = value
			continue
		}
		// Error markers pass through unchanged (per-loader error boundary)
		if isLoaderError(value) {
			result[key] = value
			continue
		}
		result[key] = pruneValue(value, fields)
	}
	return result
}

func pruneValue(value any, fields []string) any {
	var arrayFields []string
	var plainFields []string

	for _, f := range fields {
		if f == "$" {
			// Standalone $ = keep entire array elements
			return value
		}
		if len(f) > 2 && f[:2] == "$." {
			arrayFields = append(arrayFields, f[2:])
		} else {
			plainFields = append(plainFields, f)
		}
	}

	if len(arrayFields) > 0 {
		if arr, ok := value.([]any); ok {
			pruned := make([]any, len(arr))
			for i, item := range arr {
				if m, ok := item.(map[string]any); ok {
					pruned[i] = pickFields(m, arrayFields)
				} else {
					pruned[i] = item
				}
			}
			return pruned
		}
	}

	if len(plainFields) > 0 {
		if m, ok := value.(map[string]any); ok {
			return pickFields(m, plainFields)
		}
	}

	return value
}

// pickFields extracts only the listed fields from a map.
// Supports dot-separated nested paths.
func pickFields(source map[string]any, fields []string) map[string]any {
	result := make(map[string]any)
	for _, field := range fields {
		val := getNestedField(source, field)
		if val != nil {
			setNestedField(result, field, val)
		}
	}
	return result
}

func getNestedField(source map[string]any, path string) any {
	parts := splitDot(path)
	var current any = source
	for _, part := range parts {
		m, ok := current.(map[string]any)
		if !ok {
			return nil
		}
		current, ok = m[part]
		if !ok {
			return nil
		}
	}
	return current
}

func setNestedField(target map[string]any, path string, value any) {
	parts := splitDot(path)
	current := target
	for i := 0; i < len(parts)-1; i++ {
		key := parts[i]
		if next, ok := current[key]; ok {
			if nextMap, ok := next.(map[string]any); ok {
				current = nextMap
				continue
			}
		}
		next := make(map[string]any)
		current[key] = next
		current = next
	}
	current[parts[len(parts)-1]] = value
}

// isLoaderError checks if a value is a per-loader error marker.
func isLoaderError(v any) bool {
	m, ok := v.(map[string]any)
	if !ok {
		return false
	}
	errFlag, _ := m["__error"].(bool)
	if !errFlag {
		return false
	}
	_, hasCode := m["code"].(string)
	_, hasMsg := m["message"].(string)
	return hasCode && hasMsg
}

// splitDot splits a string by '.' without allocating for simple cases.
func splitDot(s string) []string {
	for i := range s {
		if s[i] != '.' {
			continue
		}
		// Has dots, use full split
		var parts []string
		start := 0
		for j := range s {
			if s[j] == '.' {
				parts = append(parts, s[start:j])
				start = j + 1
			}
		}
		parts = append(parts, s[start:])
		return parts
	}
	return []string{s}
}

File: src/server/core/typescript/src/page/projection.ts
/* src/server/core/typescript/src/page/projection.ts */

import { isLoaderError } from './loader-error.js'

export type ProjectionMap = Record<string, string[]>

const UNSAFE_PATH_SEGMENTS = new Set(['__proto__', 'prototype', 'constructor'])

function isUnsafePathSegment(part: string): boolean {
	return UNSAFE_PATH_SEGMENTS.has(part)
}

/** Set a nested field by dot-separated path, creating intermediate objects as needed. */
function setNestedField(target: Record<string, unknown>, path: string, value: unknown): void {
	const parts = path.split('.')
	let current: Record<string, unknown> = target
	for (let i = 0; i < parts.length - 1; i++) {
		const key = parts[i] as string
		if (isUnsafePathSegment(key)) return
		if (!(key in current) || typeof current[key] !== 'object' || current[key] === null) {
			current[key] = {}
		}
		current = current[key] as Record<string, unknown>
	}
	const lastPart = parts[parts.length - 1] as string
	if (isUnsafePathSegment(lastPart)) return
	current[lastPart] = value
}

/** Get a nested field by dot-separated path. */
function getNestedField(source: Record<string, unknown>, path: string): unknown {
	const parts = path.split('.')
	let current: unknown = source
	for (const part of parts) {
		if (current === null || current === undefined || typeof current !== 'object') {
			return undefined
		}
		current = (current as Record<string, unknown>)[part]
	}
	return current
}

/** Prune a single value according to its projected field paths. */
function pruneValue(value: unknown, fields: string[]): unknown {
	// Separate $ paths (array element fields) from plain paths
	const arrayFields: string[] = []
	const plainFields: string[] = []

	for (const f of fields) {
		if (f === '$') {
			// Standalone $ means keep entire array elements — return value as-is
			return value
		} else if (f.startsWith('$.')) {
			arrayFields.push(f.slice(2))
		} else {
			plainFields.push(f)
		}
	}

	if (arrayFields.length > 0 && Array.isArray(value)) {
		return value.map((item: unknown) => {
			if (typeof item !== 'object' || item === null) return item
			const pruned: Record<string, unknown> = {}
			for (const field of arrayFields) {
				const val = getNestedField(item as Record<string, unknown>, field)
				if (val !== undefined) {
					setNestedField(pruned, field, val)
				}
			}
			return pruned
		})
	}

	if (plainFields.length > 0 && typeof value === 'object' && value !== null) {
		const source = value as Record<string, unknown>
		const pruned: Record<string, unknown> = {}
		for (const field of plainFields) {
			const val = getNestedField(source, field)
			if (val !== undefined) {
				setNestedField(pruned, field, val)
			}
		}
		return pruned
	}

	return value
}

/** Prune data to only include projected fields. Missing projection = keep all. */
export function applyProjection(
	data: Record<string, unknown>,
	projections: ProjectionMap | undefined,
): Record<string, unknown> {
	if (!projections) return data

	const result: Record<string, unknown> = {}
	for (const [key, value] of Object.entries(data)) {
		if (isLoaderError(value)) {
			result[key] = value
			continue
		}
		const fields = projections[key]
		if (!fields) {
			result[key] = value
		} else {
			result[key] = pruneValue(value, fields)
		}
	}
	return result
}
